### PR TITLE
Update open telemetry dependencies and switch to using the built in JDK http client for reporting traces

### PR DESCRIPTION
**Pull request by SamBarker**
This PR removes the dependency on OkHttp and thus Kotlin in a fully backwards compatible fashion. It does this by replacing the `opentelemetry-exporter-sender-okhttp` (which is the OpenTelemetry default) with two separate senders. One for each transport. 

- `opentelemetry-exporter-sender-jdk` enables sending traces using `http/protobuf` which is the lowest common denominator option, as its the only method required by the OpenTelemetry specifications. 
- `opentelemetry-exporter-sender-grpc-managed-channel` enables support for sending traces using `grpc` and thus feature parity with the OkHttp sender.

In doing so it has brought in a newer version of the SemanticConventions library which has deprecated several of the used conventions. I have added the new fields where appropriate and suppressed the deprecation warnings as we need to propagate the changes to users.

### CODE REVIEW

File: pom.xml
Comment (ppatierno) [inline pom.xml:130]: why using two different versions and not aligning them?

File: pom.xml
Comment (ppatierno) [inline pom.xml:161]: I am not for using bom. Overwriting dependencies for many reasons (need for a specific version, CVE fixes and so on) would be a nightmare by using a bom. A bom could override everything itself. To override you should include other boms (where the override could be done) and so on. Tbh I would just use the same way we have been using without boms. @scholzj what's your take on this?

File: pom.xml
Comment (SamBarker) [reply pom.xml:130]: AFICT there is nothing to be gained from aligning them (happy to be proven wrong) so went with latest and greatest for each.

File: pom.xml
Comment (SamBarker) [reply pom.xml:161]: I'm not sure I understand the objection, a dependency with a version tag in it will override what ever is imported from the BOM. 

The BOM represents the open telemetry projects tested set of dependencies, thus reducing the  reducing the potential for inter module dependency issues. 

I'd also note that looking at the set of dependencies managed by the Otel bom are not all at the same version (the majority are at the version specified by the bom import but not all)

Comment (SamBarker) [reply pom.xml:161]: Its also worth mentioning that otel themselves say we [should use](https://arc.net/l/quote/uuvuaxoz) the bom.

Comment (ppatierno) [reply pom.xml:161]: What if I have to override a transient OTel dependency here in the bridge? And maybe this dependency is coming from a bom imported by the OTel bom itself? I will end to import this other bom here to override the dependency. I still think we should avoid using bom. It has been working fine until today without using boms and with any issues. We have the freedom to change/override dependencies when needed, without any issues.

Comment (scholzj) [reply pom.xml:161]: BOM screw up other (transitive) dependencies in a weird ways and make it hard to override them. I would definitely avoid it as well.

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: What is the impact on existing users?

Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: So currently users have to use a tracing backend exposing the OTLP related port with gRPC support (i.e. Jaeger backend has port 4317). Staying this way would not work, so it's a kind of braking change. The user has to expose the OTLP related port with HTTP support (i.e. Jaeger backend has port 4318).

Other question I have here is why forcing protobuf if the user would like to use JSON for example? The HTTP port also supports `http/json` as OTLP protocol, or? So if user doesn't set it, it would default to `grpc` which would be not supported by us, maybe we should "force" the user to set it, raising a warning/exception if not set? Or we could default to `http/protobuf` if not set.

Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: So I guess we should not do it this way? We should certainly not hardcode it into the code. But we probably first deprecate what we have and give users some options to migrate? It seems to be like this would essentially break at the time of upgrade.

Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: To avoid having a Kotlin dependency through okhttp we need to replace it with the JDK http client which doesn't support gRPC so I don't see a way for deprecation in this scenario. The OpenTelemetry library can use okhttp or the JDK http client but not both at same time to allow users to easily migrate.
The other way would be taking okhttp for one more release but saying to move using http/protobuf as protocol instead grpc. But this solution would not allow us to get rid of Kotlin dependency.

Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: @scholzj just to make my point clearer ... I see your point and I agree that a smooth migration with deprecation is better. To go this way we should first verify that okhttp really works when setting the OTLP protocol as `http/protobuf` instead of the default `grpc`. So if we go this way, we should hold on with this change and taking it for a future release (not the next one).

Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: Yeah, I guess if OkHttp supports both, we should stick with it for a few more releases and just instruct users to switch to the HTTP protocol as GRPC will not be supported in the future.

Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: >  To go this way we should first verify that okhttp really works when setting the OTLP protocol as http/protobuf instead of the default grpc.

I think you need to have the users control the switch as they also need to reconfigure ports, possibly their tracing backend etc.

Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: > I think you need to have the users control the switch as they also need to reconfigure ports, possibly their tracing backend etc.

The users should be already able to do the switch by setting the `OTEL_EXPORTER_OTLP_PROTOCOL` env var (to `http/protobuf`) when starting the bridge. This is valid on bare metal / vm.  With the operator in place, the user could use the `ContainerTemplate` instead to set the env var during the migration phase. When grpc support is dropped we could "hard code" the env var set by the operator itself directly when configuring the `KafkaBridge` instance.

Comment (scholzj) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:68]: > When grpc support is dropped we could "hard code" the env var set by the operator itself directly when configuring the KafkaBridge instance.

I guess defaulting would be nicer than hardcoding it. But yes, it seems like we can quite easily give them time to migrate to HTTP first and then we can drop OkHttp.

File: pom.xml
Comment (SamBarker) [reply pom.xml:130]: True, but it had backwards incompatible changes I didn't fancy reasoning about. I will align them when I remove the bom's

File: pom.xml
Comment (SamBarker) [reply pom.xml:161]: I've never noticed/realised I was having issues caused by using bom's. I'll remove them.

**Comment by SamBarker** [comment]
With fresh eyes I realised there was a way to restore GRPC support (however that does require OpenTelemetry 1.34.0+). By adding dependencies on `opentelemetry-exporter-sender-grpc-managed-channel` and `grpc-netty-shaded` we can maintain GRPC support and remove the Kotlin dependency. 

The build is now failing due to deprecations in the `SemanticAttributes` some of them are simple renames however others involve changing the value e.g. `SemanticAttributes.HTTP_URL` becomes `SemanticAttributes.URL_FULL` which means its not a backwards compatible change. 

Obviously I'm happy to push up the simple renames however I'm looking for input from the maintainers for what to do about the non backwards compatible changes. The build fails due to using the deprecated properties but changing them breaks compatibility.

**Comment by ppatierno** [comment]
> With fresh eyes I realised there was a way to restore GRPC support (however that does require OpenTelemetry 1.34.0+). By adding dependencies on opentelemetry-exporter-sender-grpc-managed-channel and grpc-netty-shaded we can maintain GRPC support and remove the Kotlin dependency.

Is it a documented way or an hack? Also I think that the deprecation issues makes things more complicated.

**Comment by SamBarker** [comment]
Oh and as I forgot to mention it earlier the need for 1.34.0 rather than 1.28.0 where the sender was introduced was [this bug fix](https://github.com/open-telemetry/opentelemetry-java/pull/6110)

File: pom.xml
Comment (SamBarker) [reply pom.xml]: bom's removed in https://github.com/strimzi/strimzi-kafka-bridge/pull/867/commits/25df0af1df7f7ee1ff530e3a91e376f612c895cc

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (SamBarker) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java]: `url.full` is a client specified attribute the [server side specification](https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-server-semantic-conventions) breaks it up into component parts. However `url.full` is the documented replacement for `http.url`. So I'm taking the opportunity of deprecating `http.url` to improve the spec compliance.

File: pom.xml
Comment (ppatierno) [reply pom.xml:131]: About using 2.0 we have to pay more attention.
It seems there are a lot of breaking changes https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v2.0.0
The main ones impacting us would be:

* defaulting the OTLP protocol from `grpc` to `http/protobuf`. To keep backward compatibility on our side, it means we should "force" grpc in the bridge `otel.exporter.otlp.protocol=grpc` if the user doesn't set `OTEL_EXPORTER_OTLP_PROTOCOL` env var explicitely.
* we should explicitely disable the logs exporter `otel.logs.exporter=none` as we do for the metrics exporter already.

File: pom.xml
Comment (ppatierno) [inline pom.xml:142]: can we move this definition together with all the others opentelemetry related above? Also it seems latest release is [1.23.1-alpha](https://github.com/open-telemetry/semantic-conventions-java/releases/tag/v1.23.1)

File: documentation/modules/proc-configuring-kafka-bridge-tracing.adoc
Comment (ppatierno) [inline documentation/modules/proc-configuring-kafka-bridge-tracing.adoc:98]: I haven't had a section like this in the past. I am not sure we want it now otherwise we should list all span attributes (i.e. otel.status_code, span.kind, ....). I would take them out but of course taking what you have in the CHANGELOG mentioning new attributes and the ones deprecated and planned to be removed.

File: CHANGELOG.md
Comment (ppatierno) [inline CHANGELOG.md:15]: why it's mentioned twice? I mean the attribute name and then the same with the value which in our case will always be `topic`.

File: pom.xml
Comment (ppatierno) [inline pom.xml:269]: out of curiosity, why this one needs to be excluded?

File: pom.xml
Comment (ppatierno) [inline pom.xml:287]: can you elaborate why this dependency is needed please?

File: documentation/modules/proc-configuring-kafka-bridge-tracing.adoc
Comment (SamBarker) [reply documentation/modules/proc-configuring-kafka-bridge-tracing.adoc:98]: I added them as it felt a bit random to just document them in the changelog with out mentioning them anywhere else. As much for discoverability as anything else. 

I only documented the bridge specific tags, which seems like a reasonable choice. We could add a sentence along the lines of `In addition to the open telemetry standard tags we include the following`.

File: CHANGELOG.md
Comment (SamBarker) [reply CHANGELOG.md:15]: I phrased it clumsily I was trying to capture that both the key and value types were being deprecated.

File: pom.xml
Comment (SamBarker) [reply pom.xml:287]: This is the dependency which provides the grpc transport that the GRPC sender uses instead of OkHttp.

File: pom.xml
Comment (SamBarker) [reply pom.xml:269]: Hmm good question, I copied from the infinispan change I'll dig a bit deeper.

File: pom.xml
Comment (SamBarker) [reply pom.xml:131]: I can bump to 2.0 if you would prefer.

File: documentation/modules/proc-configuring-kafka-bridge-tracing.adoc
Comment (ppatierno) [inline documentation/modules/proc-configuring-kafka-bridge-tracing.adoc:99]:
```suggestion

The Kafka Bridge adds, in addition to the standard OpenTelemetry attributes, the following attributes from the https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-server-semantic-conventions[OpenTelemetry standard conventions for HTTP] to its spans.
```

**Review by ppatierno — COMMENTED** [review]
@SamBarker I had another pass but I see some comments were not addressed.

File: pom.xml
Comment (SamBarker) [reply pom.xml:269]: I've removed the exclusion and it doesn't make a difference to the `dependency:tree` so I guess it doesn't apply in our context.

File: pom.xml
Comment (SamBarker) [reply pom.xml:131]: Do you want to move to 2.0 now or not? I've addressed the logs exporter point as that applied without moving to 2.0.0.

Comment (ppatierno) [reply pom.xml:131]: I would avoid moving to a 2.x version for now. It would be anyway only for the instrumented Kafka client but I would leave using a 1.3x version for now.

File: CHANGELOG.md
Comment (ppatierno) [inline CHANGELOG.md:8]: I guess we are missing all the other bumps you made in the pom.
Also it's enough just naming component - version so for example ...

```shell
Dependency update (OpenTelemetry Semconv 1.21.0-alpha, ....)
```

Comment (scholzj) [reply CHANGELOG.md:8]: This probably should also not have its own line in the changelog but be added to the existing dependency updates?

Comment (ppatierno) [reply CHANGELOG.md:8]: Yeah right.

Comment (SamBarker) [reply CHANGELOG.md:8]: I've added Semconv to the global dependency list, but thought it was worth calling out this particular change as it leads to the changes of the tags included on spans. 

I can re-phrase if that is desired but I think mentioning the changes is merited.

Comment (scholzj) [reply CHANGELOG.md:8]: If it is for whatever reason more important, you should probably rephrase it then:
* It doesn't make sense to have to separate entries for `Dependency update (...)`
* If this is more important for some reason, explain it there. If it changes some tags, make it clear there what and how.

You seem to have the list of changes already, so you can just say something like 
```
* OpenTelemetry Semconv update causes following incompatible changes:
    * ...
```

File: CHANGELOG.md
Comment (scholzj) [inline CHANGELOG.md:9]: `are being replaced replacements` seems like a typo.

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:54]: The CHANGELOG suggests these are removed. Here you say remove after release 0.28. Why? If we keep them for backwards compatibility, that is great. But might make sense to give it few more releases. It should be also explained in the changelog properly what is being done.

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:73]: MAybe the comment can add some more explanation what logs and why are we disabling them?

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:98]: A comment what is deprecated might be useful. Also, if something should be removed later, an issue would be nice for tracking and link it in the comments.

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java]: I've updated the changelog to better reflect the change and reworded this comment to more accurately reflect the intent.

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java]: I've added a comment explaining the suppressions and raised https://github.com/strimzi/strimzi-kafka-bridge/issues/875 to remove the deprecated attributes.

File: CHANGELOG.md
Comment (ppatierno) [inline CHANGELOG.md:8]:
```suggestion
* This release deprecates several attributes (inline with changes from OpenTelemetry Semconv) which it attaches to spans. Both the deprecated attribute and its replacement will be added to spans in the current release. The deprecated attributes will be removed in a future version.
```

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:73]: I know it's not related to this PR but you expand the comment about "disable metrics" the same way for consistency please? Something like you have for logs, i.e. `// disable otel metrics exporter`. It sounds useless opening a different PR just for that. Thanks.

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java]: Resolved with https://github.com/strimzi/strimzi-kafka-bridge/pull/867/commits/8474fbb6d2b0cdbca3bee4237ca340a5702ee357

File: src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java:73]: resolved with https://github.com/strimzi/strimzi-kafka-bridge/pull/867/commits/14e150d7a0104c235e2094a57afec185d535c19d

**Comment by SamBarker** [comment]
@ppatierno @scholzj I believe everything has been addressed and the PR checks are happy modulo a PR approval  😉

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks for the PR!

**Comment by ppatierno** [comment]
I tested (again) this PR and it seems to work fine, even switching between using `grpc` (default) to `http/protobuf`.
@SamBarker could you fix the conflicts please? We want to validate that with the same versions (as you set on https://github.com/strimzi/strimzi-kafka-operator/pull/9758), OpenTelemetry works in the operator as well (KafkaConnect and KafkaMirrorMaker). After this check, I think we can move forward and merge both.

**Comment by SamBarker** [comment]
> We want to validate that with the same versions (as you set on [strimzi/strimzi-kafka-operator#9758](https://github.com/strimzi/strimzi-kafka-operator/pull/9758)), OpenTelemetry works in the operator as well (KafkaConnect and KafkaMirrorMaker). After this check, I think we can move forward and merge both.

I'm not clear what you want me to do, if anything, for that. The versions are the same across both this and the operator PR and the operator regression suite passed. Though that ran with the version of the bridge with out these changes.

**Review by scholzj — APPROVED** [review]

**Comment by ppatierno** [comment]
> I'm not clear what you want me to do, if anything, for that. The versions are the same across both this and the operator PR and the operator regression suite passed. Though that ran with the version of the bridge with out these changes.

Maybe it was a misleading comment but I didn't want nothing to do from you :-D
I meant to wait for the PR on the operator to work fine before merging this one, in order to validate that same OTel versions were working fine on both (operator and bridge).

**Pull request merged**
